modules/renter/contractor/host_integration_test.go

```diff
@@ -573,3 +574,103 @@ func TestIntegrationRenew(t *testing.T) {
 		t.Fatal(contract.FileContract.WindowStart)
 	}
 }
+
+// TestResync tests that the contractor can resync with a host after being
+// interrupted during contract revision.
+func TestResync(t *testing.T) {
+	if testing.Short() {
+		t.SkipNow()
+	}
+	t.Parallel()
+	// create testing trio
+	h, c, _, err := newTestingTrio("TestResync")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// get the host's entry from the db
+	hostEntry, ok := c.hdb.Host(h.ExternalSettings().NetAddress)
+	if !ok {
+		t.Fatal("no entry for host in db")
+	}
+
+	// form a contract with the host
+	contract, err := c.managedNewContract(hostEntry, 10, c.blockHeight+100)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// revise the contract
+	editor, err := c.Editor(contract)
+	if err != nil {
+		t.Fatal(err)
+	}
+	data, err := crypto.RandBytes(int(modules.SectorSize))
+	if err != nil {
+		t.Fatal(err)
+	}
+	root, err := editor.Upload(data)
+	if err != nil {
+		t.Fatal(err)
+	}
+	err = editor.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// download the data
+	contract = c.contracts[contract.ID]
+	downloader, err := c.Downloader(contract)
+	if err != nil {
+		t.Fatal(err)
+	}
+	retrieved, err := downloader.Sector(root)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !bytes.Equal(data, retrieved) {
+		t.Fatal("downloaded data does not match original")
+	}
+	err = downloader.Close()
+	if err != nil {
+		t.Fatal(err)
+	}
+	contract = c.contracts[contract.ID]
+
+	// corrupt contract and delete its cachedRevision
+	badContract := contract
+	badContract.LastRevision.NewRevisionNumber--
+	badContract.LastRevisionTxn.TransactionSignatures = nil // delete signatures
+
+	c.mu.Lock()
+	delete(c.cachedRevisions, contract.ID)
+	c.mu.Unlock()
+
+	// Editor and Downloader should fail with the bad contract
+	_, err = c.Editor(badContract)
+	if !proto.IsRevisionMismatch(err) {
+		t.Fatal("expected revision mismatch, got", err)
+	}
+	_, err = c.Downloader(badContract)
+	if !proto.IsRevisionMismatch(err) {
+		t.Fatal("expected revision mismatch, got", err)
+	}
+
+	// add cachedRevision
+	c.mu.Lock()
+	c.cachedRevisions[contract.ID] = contract.LastRevision
```

Comment:
This is still an issue, because on the live network the cached revision is not going to have the signatures, but here the last revision does have the signatures.

Unless I'm missing something.